drone/exec2/exec.go

```diff
@@ -0,0 +1,366 @@
+	envs := environ.Combine(
+		getEnv(cliContext),
+		environ.System(commy.System),
+		environ.Repo(commy.Repo),
+		environ.Build(commy.Build),
+		environ.Stage(commy.Stage),
+		environ.Link(commy.Repo, commy.Build, commy.System),
+		commy.Build.Params,
+	)
+
+	// string substitution function ensures that string
+	// replacement variables are escaped and quoted if they
+	// contain newlines.
+	subf := func(k string) string {
+		v := envs[k]
+		if strings.Contains(v, "\n") {
+			v = fmt.Sprintf("%q", v)
+		}
+		return v
+	}
+
+	// evaluates string replacement expressions and returns an
+	// update configuration.
+	config, err := envsubst.Eval(string(rawsource), subf)
+	if err != nil {
+		return err
+	}
+
+	// parse and lint the configuration.
+	manifest, err := manifest.ParseString(config)
+	if err != nil {
+		return err
+	}
+
+	// a configuration can contain multiple pipelines.
+	// get a specific pipeline resource for execution.
+	if commy.Stage.Name == "" {
+		fmt.Println("No stage specified, assuming 'default'")
+	}
+
+	res, err := resource.Lookup(commy.Stage.Name, manifest)
+	if err != nil {
+		return fmt.Errorf("Stage '%s' not found in build file : %s", commy.Stage.Name, err)
+	}
+
+	// lint the pipeline and return an error if any
+	// linting rules are broken
+	lint := linter.New()
+	err = lint.Lint(res, commy.Repo)
+	if err != nil {
+		return err
+	}
+
+	// compile the pipeline to an intermediate representation.
+	comp := &compiler.Compiler{
+		Environ:    provider.Static(commy.Environ),
+		Labels:     commy.Labels,
+		Resources:  commy.Resources,
+		Tmate:      commy.Tmate,
+		Privileged: append(commy.Privileged, compiler.Privileged...),
+		Networks:   commy.Networks,
+		Volumes:    commy.Volumes,
+		Secret:     secret.StaticVars(commy.Secrets),
+		Registry: registry.Combine(
+			registry.File(commy.Config),
+		),
+	}
+
+	// when running a build locally cloning is always
+	// disabled in favor of mounting the source code
+	// from the current working directory.
+	if !commy.Clone {
+		comp.Mount, _ = os.Getwd()
+	}
+
+	args := runtime.CompilerArgs{
+		Pipeline: res,
+		Manifest: manifest,
+		Build:    commy.Build,
+		Netrc:    commy.Netrc,
+		Repo:     commy.Repo,
+		Stage:    commy.Stage,
+		System:   commy.System,
+	}
+	spec := comp.Compile(nocontext, args).(*engine.Spec)
+
+	// include only steps that are in the include list,
+	// if the list in non-empty.
+	if len(commy.Include) > 0 {
+	I:
+		for _, step := range spec.Steps {
+			if step.Name == "clone" {
+				continue
+			}
+			for _, name := range commy.Include {
+				if step.Name == name {
+					continue I
+				}
+			}
+			step.RunPolicy = runtime.RunNever
+		}
+	}
+	// exclude steps that are in the exclude list, if the list in non-empty.
+	if len(commy.Exclude) > 0 {
+	E:
+		for _, step := range spec.Steps {
+			if step.Name == "clone" {
+				continue
+			}
+			for _, name := range commy.Exclude {
+				if step.Name == name {
+					step.RunPolicy = runtime.RunNever
+					continue E
```

Comment:
you can simplify this simply by using `break` instead of `continue E`...

Reply:
this is copied verbatim from the docker runner, and i hesitate to change the logic.